# Task 2 - to predict scores of  a student based on the no. of hours he studies.
# using supervised ML

In [14]:
import pandas as pd
import numpy as np
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_absolute_error
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
import matplotlib.pyplot as plt

stud_filepath = 'https://raw.githubusercontent.com/AdiPersonalWorks/Random/master/student_scores%20-%20student_scores.csv'
stud_data = pd.read_csv(stud_filepath)
stud_data.describe()

Hours     Scores
count  25.000000  25.000000
mean    5.012000  51.480000
std     2.525094  25.286887
min     1.100000  17.000000
25%     2.700000  30.000000
50%     4.800000  47.000000
75%     7.400000  75.000000
max     9.200000  95.000000

In [2]:
stud_data.columns
stud_data = stud_data.dropna(axis=0)
y = stud_data.Scores
stud_features = ['Hours']
X = stud_data[stud_features]
X.describe()


Hours
count  25.000000
mean    5.012000
std     2.525094
min     1.100000
25%     2.700000
50%     4.800000
75%     7.400000
max     9.200000

In [3]:
stud_model = DecisionTreeRegressor(random_state=1)
stud_model.fit(X, y)
print('Predicted student Scores =', stud_model.predict(X))

Predicted student Scores = [25.5 47.  27.  75.  30.  20.  88.  60.  81.  27.5 85.  62.  41.  42.
 17.  95.  25.5 24.  67.  69.  27.5 54.  35.  76.  86. ]


## Predicted scores are exactly the same as the actual scores of the dataset (except for 2 values)
### hours = 2.5   actual_score = 21    predicted_score = 25.5
### hours = 2.7   actual_score = 30    predicted_score = 27.5
## this model is bias because
### hours = 2.5   actual_score = 21    predicted_score = 25.5
### hours = 2.5   actual_score = 30    predicted_score = 25.5
### hours = 2.7   actual_score = 25    predicted_score = 27.5
### hours = 2.7   actual_score = 30    predicted_score = 27.5

In [4]:
predicted_stud_scores = stud_model.predict(X)
print(mean_absolute_error(y, predicted_stud_scores))

0.56


#### MAE = 0.56 which is almost null for the given dataset because the data is bias to a very small extent

## Splitting the dataset into training data and validation data

In [5]:
train_X, val_X, train_y, val_y = train_test_split(X, y, random_state=0)
stud_model = DecisionTreeRegressor()
stud_model.fit(train_X, train_y)
val_predictions = stud_model.predict(val_X)
print(mean_absolute_error(val_y, val_predictions))

8.0


### after splitting the data into training data and validation data, MAE has increased to 8

In [6]:
def get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y):
    model = DecisionTreeRegressor(max_leaf_nodes = max_leaf_nodes, random_state=0)
    model.fit(train_X, train_y)
    pred_val = model.predict(val_X)
    mae = mean_absolute_error(val_y, pred_val)
    return mae

for max_leaf_nodes in [2, 4, 6, 8 ,10]:
    my_mae = get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y)
    print("Max leaf nodes: %d   \t\t    Mean Absolute Error: %d" %(max_leaf_nodes, my_mae))


Max leaf nodes: 2   		    Mean Absolute Error: 9
Max leaf nodes: 4   		    Mean Absolute Error: 7
Max leaf nodes: 6   		    Mean Absolute Error: 5
Max leaf nodes: 8   		    Mean Absolute Error: 6
Max leaf nodes: 10   		    Mean Absolute Error: 7


### from the above options, 6 is the optimal no. of leaves, which has the least MAE of 5
## Underfitting and Overfitting

In [7]:
forest_model = RandomForestRegressor(random_state=0)
forest_model.fit(train_X, train_y)
stud_forest_pred = forest_model.predict(val_X)
print(mean_absolute_error(val_y, stud_forest_pred))

5.231309523809524


### Using random forest tree, I got MAE = 5.2 
### and after finding the sweet spot between underfitting and overfitting, with 6 leaf nodes, I got MAE = 5


# What will be predicted score if a student study for 9.25 hrs in a day?

In [19]:
hours = 9.25
p = forest_model.predict([[hours]])
print('no. of hours = ', hours)
print('predicted score =', p[0])

no. of hours =  9.25
predicted score = 88.92


### I conclude that the scores can be best predicted using 6 leaf nodes, which gives the least MAE = 5
### To predict accurate scores when the dataset becomes little bigger, we can use Random Forest Regessor model.